### ISSUE TITLE
Generic.WhiteSpace.ScopeIndent - indent in closure

### ISSUE THREAD

**Issue by webimpress** [squizlabs/PHP_CodeSniffer#1535]
I want to align object operators in fluent calls:

```php
        $mockSendResponseEvent->expects($this->any())
                              ->method('contentSent')
                              ->will($this->returnCallback(function () use (&$returnValue) {
                                  if (false === $returnValue) { // <!-- error here
                                      $returnValue = true;
                                      return false;
                                  } // <-- error here
                                  return true;
                              }));
```

The callback body should be then aligned one more indent (4 spaces) than parent line, so it should be as the example above. But in lines marked above I got following errors:
```
Line indented incorrectly; expected 32 spaces, found 34 (Generic.WhiteSpace.ScopeIndent.IncorrectExact)
```

Please note that I have got errors only on these lines, all others are fine.

**Comment by gsherwood** [comment]
This is because your code is not aligned with tab stops properly. If you tabbed the code in your text editor, the IF should be either aligned under the `w` in `will(` or under the `(`, which would be 32 or 36 spaces. 30 spaces isn't a valid tab stop for a 4-space indented file, so the sniff complains.

In this specific case, the sniffs wants you to align your code at tab stop 8 (32 spaces), so it looks like this (not I've removed indent from the left side for simplicity but the alignment is the same):
```php
<?php

$mockSendResponseEvent->expects($this->any())
                      ->method('contentSent')
                      ->will($this->returnCallback(function () use (&$returnValue) {
                        if (false === $returnValue) {
                            $returnValue = true;
                            return false;
                        }
                        return true;
                      }));

```
The scope indent sniff does not have any rules in it that allows it to know that you want to do mid-stop indenting, and only for when you are trying to align object operators. I can't see how I would add that either as it can't assume everything is trying to align object operators in this way. So I don't think I can solve this problem for you.

The only thing I can suggest is the other method of aligning object operators that I see people use. That would make your code looks like this (no mid-stop indents used):
```php
$mockSendResponseEvent
    ->expects($this->any())
    ->method('contentSent')
    ->will($this->returnCallback(function () use (&$returnValue) {
        if (false === $returnValue) {
            $returnValue = true;
            return false;
        }
        return true;
    }));
```
I believe this formatting method is fairly common because it is easy to achieve with tabbing. It also works for both tab and space indented files.

**Comment by webimpress** [comment]
@gsherwood Thanks for your answer.
In general I like code formatted as you suggested, with first call starting in new line, but in case:
```php
$this
    ->method1()
    ->method2(function () {
        if ($something) {
            return 1;
        }
        return 2;
    });
```
it looks a bit weird and I would prefer:
```php
$this->method1()
     ->method2(function () {
         if ($something) {
             return 1;
         }
         return 2;
     });
```
Here we will get the same issue with indent of `if` statement. We have there 9 spaces before `if` and  IMHO it should pass PSR-2 tests (body of closure is indented once). Maybe previously it wasn't clear - but why we get errors only on incorrect indents of `if` statement and no any errors on other lines - for example with `return`?

**Comment by gsherwood** [comment]
> Maybe previously it wasn't clear - but why we get errors only on incorrect indents of if statement and no any errors on other lines - for example with return?

The reason is because lines that open a block of code must be indented exactly to the required column while (by default) lines that do not open blocks must be indent to *at least* the required column. This lets you define your own indent rules within code blocks (like custom array and comment alignment rules) while still maintaining consistency of the overall structure.

So in the ScopeIndent sniff, the code blocks define the indent rules and so they must also adhere to them exactly. If they don't, it can throw out the indenting of the entire file.

I hope that explains things better,

**Comment by webimpress** [comment]
@gsherwood thanks for your answer.

If I understand you correctly you suggest that I can define my own indentation rules for code block, i.e. closures? If so I think it will be pretty hard (impossible?) with enabled `Generic.WhiteSpace.ScopeIndent`.
For example if I would define different indentation (let say 1 space indent) for closures which contain (let say) `if` statement - it will be impossible, because the `if` statement MUST be indented to the global skeleton (4 spaces indent).

Is it possible to disable in `Generic.WhiteSpace.ScopeIndent` checking the whole body of the code block?

**Comment by gsherwood** [comment]
> If I understand you correctly you suggest that I can define my own indentation rules for code block, i.e. closures?

No. Closures open scope and so define indents. You can indent lines within the closure as far as you want, as long as they don't also open scopes and code blocks.

> Is it possible to disable in Generic.WhiteSpace.ScopeIndent checking the whole body of the code block?

No
